File: threedollar-api-user-service/src/main/java/com/depromeet/threedollar/api/user/service/auth/GoogleAuthService.java
package com.depromeet.threedollar.api.user.service.auth;

import org.springframework.stereotype.Service;

import com.depromeet.threedollar.api.user.service.auth.dto.request.LoginRequest;
import com.depromeet.threedollar.api.user.service.auth.dto.request.SignUpRequest;
import com.depromeet.threedollar.api.user.service.user.UserService;
import com.depromeet.threedollar.api.user.service.user.UserServiceUtils;
import com.depromeet.threedollar.common.utils.HttpHeaderUtils;
import com.depromeet.threedollar.domain.rds.domain.userservice.user.UserRepository;
import com.depromeet.threedollar.domain.rds.domain.userservice.user.UserSocialType;
import com.depromeet.threedollar.external.client.google.GoogleAuthApiClient;
import com.depromeet.threedollar.external.client.google.dto.response.GoogleProfileInfoResponse;

import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
@Service
class GoogleAuthService implements AuthService {

    private static final UserSocialType socialType = UserSocialType.GOOGLE;

    private final GoogleAuthApiClient googleAuthApiClient;
    private final UserService userService;
    private final UserRepository userRepository;

    @Override
    public Long signUp(SignUpRequest request) {
        GoogleProfileInfoResponse response = googleAuthApiClient.getProfileInfo((HttpHeaderUtils.withBearerToken(request.getToken())));
        return userService.registerUser(request.toCreateUserRequest(response.getId()));
    }

    @Override
    public Long login(LoginRequest request) {
        GoogleProfileInfoResponse response = googleAuthApiClient.getProfileInfo((HttpHeaderUtils.withBearerToken(request.getToken())));
        return UserServiceUtils.findUserBySocialIdAndSocialType(userRepository, response.getId(), socialType).getId();
    }

}

File: threedollar-api-admin/src/main/kotlin/com/depromeet/threedollar/api/admin/service/auth/GoogleAuthService.kt
package com.depromeet.threedollar.api.admin.service.auth

import com.depromeet.threedollar.api.admin.service.auth.dto.request.LoginRequest
import com.depromeet.threedollar.common.exception.model.NotFoundException
import com.depromeet.threedollar.common.exception.type.ErrorCode
import com.depromeet.threedollar.common.utils.HttpHeaderUtils
import com.depromeet.threedollar.domain.rds.user.domain.admin.AdminRepository
import com.depromeet.threedollar.external.client.google.GoogleAuthApiClient
import org.springframework.stereotype.Service

@Service
class GoogleAuthService(
    private val adminRepository: AdminRepository,
    private val googleAuthApiClient: GoogleAuthApiClient
) : AuthService {

    override fun login(request: LoginRequest): Long {
        val googleProfileInfo = googleAuthApiClient.getProfileInfo(HttpHeaderUtils.withBearerToken(request.token))
        val admin = adminRepository.findAdminByEmail(googleProfileInfo.email)
            ?: throw NotFoundException("해당하는 관리자 (${googleProfileInfo.email})은 존재하지 않습니다", ErrorCode.NOTFOUND_ADMIN)
        return admin.id
    }

}
